web/js/precallView.js

```diff
@@ -0,0 +1,236 @@
+/* globals setTimeout */
+!(function (exports) {
+  'use strict';
+
+  var _templateSrc = '/templates/precall.ejs';
+  var _template;
+  var _model;
+  var testMeterInterval;
+
+  var addHandlers = function () {
+    var preCallTestResults = document.getElementById('pre-call-test-results');
+
+    preCallTestResults.addEventListener('click', function (e) {
+      var elem = e.target;
+      switch (elem.id) {
+        case 'precall-close':
+          preCallTestResults.style.display = 'none';
+          break;
+        case 'retest':
+          preCallTestResults.style.display = 'none';
+          Utils.sendEvent('roomView:retest');
+          break;
+      }
+    });
+
+    var publishSettings = document.querySelector('.publish-settings');
+
+    publishSettings.addEventListener('click', function (e) {
+      var initialVideoSwitch = document.querySelector('#initialVideoSwitch');
+      var initialAudioSwitch = document.querySelector('#initialAudioSwitch');
+      var elem = e.target;
+      elem.blur();
+      // pointer-events is not working on IE so we can receive as target a child
+      elem = HTMLElems.getAncestorByTagName(elem, 'a');
+      if (!elem) {
+        return;
+      }
+      switch (elem.id) {
+        case 'initialAudioSwitch':
+          if (!initialAudioSwitch.classList.contains('activated')) {
+            setSwitchStatus(true, true, 'Audio', 'roomView:initialAudioSwitch');
+          } else {
+            setSwitchStatus(false, true, 'Audio', 'roomView:initialAudioSwitch');
+          }
+          break;
+        case 'initialVideoSwitch':
+          if (!initialVideoSwitch.classList.contains('activated')) {
+            setSwitchStatus(true, true, 'Video', 'roomView:initialVideoSwitch');
+          } else {
+            setSwitchStatus(false, true, 'Video', 'roomView:initialVideoSwitch');
+          }
+          break;
+      }
+    });
+
+    var videoPreviewElement = document.getElementById('video-preview');
+    var videoPreviewNameElement = document.getElementById('video-preview-name');
+    videoPreviewElement.addEventListener('mouseover', function () {
+      videoPreviewNameElement.style.opacity = 1;
+    });
+    videoPreviewElement.addEventListener('mouseout', function () {
+      videoPreviewNameElement.style.opacity = 0;
+    });
+  };
+
+  function render(resolve) {
+    _template.render().then(function (aHTML) {
+      document.body.innerHTML += aHTML;
+      addHandlers();
+      resolve();
+    });
+  }
+
+  var eventHandlers = {
+    'PrecallController:endPrecall': function () {
+      _model.addEventListener('value', render);
+      render();
+    },
+    'PrecallController:audioOnly': function () {
+      setSwitchStatus(false, true, 'Video', 'roomView:initialVideoSwitch');
```

Comment:
Is there a way we could make this function clearer? I'm not a fan of listing booleans and having to get check what each one does.

Reply:
I've decided to pull the second Boolean parameter. The event can always bubble up.

Reply:
Perfect 👍